### PR TITLE
Refactor plugin loader

### PR DESCRIPTION
**Pull request by dachuan9e**
Changes proposed in this pull request:
- Refactor plugin loader

### CODE REVIEW

File: shardingsphere-agent/shardingsphere-agent-core/src/main/java/org/apache/shardingsphere/agent/core/plugin/Loader.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.shardingsphere.agent.core.plugin;
+
+import net.bytebuddy.description.type.TypeDescription;
+import org.apache.shardingsphere.agent.api.point.PluginInterceptorPoint;
+
+public interface Loader {
```
Comment (terrymanu) [inline shardingsphere-agent/shardingsphere-agent-core/src/main/java/org/apache/shardingsphere/agent/core/plugin/Loader.java:23]: `Loader` is too generic, `PluginLoader` is better

File: shardingsphere-agent/shardingsphere-agent-core/src/main/java/org/apache/shardingsphere/agent/core/plugin/Loader.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+     *
+     * @param typeDescription TypeDescription
+     * @return contains when it is true
+     */
+    boolean containsType(TypeDescription typeDescription);
+    
+    /**
+     * Load plugin interceptor point by TypeDescription.
+     *
+     * @param typeDescription TypeDescription
```
Comment (terrymanu) [inline shardingsphere-agent/shardingsphere-agent-core/src/main/java/org/apache/shardingsphere/agent/core/plugin/Loader.java:35]: `@param typeDescription TypeDescription` should rewrite to `@param typeDescription type description`

File: shardingsphere-agent/shardingsphere-agent-core/src/main/java/org/apache/shardingsphere/agent/core/plugin/Loader.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    PluginInterceptorPoint loadPluginInterceptorPoint(TypeDescription typeDescription);
+    
+    /**
+     * To get or create instance of the advice class. Create new one and caching when it is not exist.
+     *
+     * @param classNameOfAdvice class name of advice
+     * @param <T> advice type
+     * @return instance of advice
+     */
+    <T> T getOrCreateInstance(String classNameOfAdvice);
```
Comment (terrymanu) [inline shardingsphere-agent/shardingsphere-agent-core/src/main/java/org/apache/shardingsphere/agent/core/plugin/Loader.java:47]: It is better to rename `classNameOfAdvice` to `adviceClassName`

**Review by terrymanu — CHANGES_REQUESTED** [review]

File: shardingsphere-agent/shardingsphere-agent-core/src/main/java/org/apache/shardingsphere/agent/core/bytebuddy/transformer/ShardingSphereTransformer.java [hunk 4]
```diff
@@ -112,27 +113,13 @@
...
-                        return new ShardingSphereTransformationPoint<>(md, new ClassStaticMethodAroundInterceptor(pluginLoader.getOrCreateInstance(classStaticMethodPoints.get(0).getAdvice())));
-                    } else {
-                        Collection<ClassStaticMethodAroundAdvice> classStaticMethodAroundAdvices = classStaticMethodPoints.stream()
-                                .map(ClassStaticMethodPoint::getAdvice)
-                                .map(advice -> (ClassStaticMethodAroundAdvice) pluginLoader.getOrCreateInstance(advice))
-                                .collect(Collectors.toList());
-                        return new ShardingSphereTransformationPoint<>(md, new ComposeClassStaticMethodAroundInterceptor(classStaticMethodAroundAdvices));
-                    }
-                })
+                .map(md -> getStaticClassTransformationPoint(md, classStaticMethodAroundPoints))
```
Comment (terrymanu) [inline shardingsphere-agent/shardingsphere-agent-core/src/main/java/org/apache/shardingsphere/agent/core/bytebuddy/transformer/ShardingSphereTransformer.java:118]: What is `md`? Is it better to change another name?

File: shardingsphere-agent/shardingsphere-agent-core/src/main/java/org/apache/shardingsphere/agent/core/plugin/Loader.java (hunk 3 above)
Comment (dachuan9e) [reply shardingsphere-agent/shardingsphere-agent-core/src/main/java/org/apache/shardingsphere/agent/core/plugin/Loader.java:47]: it's better.

File: shardingsphere-agent/shardingsphere-agent-core/src/main/java/org/apache/shardingsphere/agent/core/plugin/Loader.java (hunk 2 above)
Comment (dachuan9e) [reply shardingsphere-agent/shardingsphere-agent-core/src/main/java/org/apache/shardingsphere/agent/core/plugin/Loader.java:35]: ok, thanks

File: shardingsphere-agent/shardingsphere-agent-core/src/main/java/org/apache/shardingsphere/agent/core/bytebuddy/transformer/ShardingSphereTransformer.java (hunk 4 above)
Comment (dachuan9e) [reply shardingsphere-agent/shardingsphere-agent-core/src/main/java/org/apache/shardingsphere/agent/core/bytebuddy/transformer/ShardingSphereTransformer.java]: It's method description

File: shardingsphere-agent/shardingsphere-agent-core/src/main/java/org/apache/shardingsphere/agent/core/bytebuddy/transformer/ShardingSphereTransformer.java [hunk 5]
```diff
@@ -145,28 +132,32 @@
...
+        if (classStaticMethodPoints.size() == 1) {
+            return new ShardingSphereTransformationPoint<>(md, new ClassStaticMethodAroundInterceptor(pluginLoader.getOrCreateInstance(classStaticMethodPoints.get(0).getAdvice())));
+        } else {
+            Collection<ClassStaticMethodAroundAdvice> classStaticMethodAroundAdvices = new LinkedList<>();
+            for (ClassStaticMethodPoint point : classStaticMethodPoints) {
+                if (null != point.getAdvice()) {
+                    classStaticMethodAroundAdvices.add(pluginLoader.getOrCreateInstance(point.getAdvice()));
+                }
+            }
+            return new ShardingSphereTransformationPoint<>(md, new ComposeClassStaticMethodAroundInterceptor(classStaticMethodAroundAdvices));
```
Comment (terrymanu) [inline shardingsphere-agent/shardingsphere-agent-core/src/main/java/org/apache/shardingsphere/agent/core/bytebuddy/transformer/ShardingSphereTransformer.java:149]: Please rename `md` to `methodDescription` too.
It is better to do the global search

File: shardingsphere-agent/shardingsphere-agent-core/src/main/java/org/apache/shardingsphere/agent/core/bytebuddy/transformer/ShardingSphereTransformer.java [hunk 6]
```diff
@@ -178,4 +169,22 @@
         }
         return result;
     }
+    
+    private ShardingSphereTransformationPoint<?> getInstanceMethodTransformationPoint(final MethodDescription.InDefinedShape md, final List<InstanceMethodPoint> instanceMethodAroundPoints) {
```
Comment (terrymanu) [inline shardingsphere-agent/shardingsphere-agent-core/src/main/java/org/apache/shardingsphere/agent/core/bytebuddy/transformer/ShardingSphereTransformer.java:173]: Please do not use inline import, just use `InDefinedShape`.
By the way, `md` is not match to `InDefinedShape`, is it?

File: shardingsphere-agent/shardingsphere-agent-core/src/main/java/org/apache/shardingsphere/agent/core/plugin/PluginApmLoader.java [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.util.jar.Manifest;
+import java.util.zip.ZipEntry;
+
+/**
+ * Plugin loader.
+ */
+@Slf4j
+public final class PluginApmLoader extends ClassLoader implements Closeable, PluginLoader {
```
Comment (terrymanu) [inline shardingsphere-agent/shardingsphere-agent-core/src/main/java/org/apache/shardingsphere/agent/core/plugin/PluginApmLoader.java:64]: `PluginApmLoader` or `ApmPluginLoader`?

File: shardingsphere-agent/shardingsphere-agent-core/src/main/java/org/apache/shardingsphere/agent/core/plugin/PluginApmLoader.java [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+        return null;
+    }
+    
+    @Override
+    public void close() {
+        for (PluginJar each : jars) {
+            try {
+                each.jarFile.close();
+            } catch (final IOException ex) {
+                log.error("close is ", ex);
```
Comment (terrymanu) [inline shardingsphere-agent/shardingsphere-agent-core/src/main/java/org/apache/shardingsphere/agent/core/plugin/PluginApmLoader.java:236]: Error message "close is " is not clear

**Review by terrymanu — CHANGES_REQUESTED** [review]

File: shardingsphere-agent/shardingsphere-agent-core/src/main/java/org/apache/shardingsphere/agent/core/bytebuddy/transformer/ShardingSphereTransformer.java (hunk 6 above)
Comment (dachuan9e) [reply shardingsphere-agent/shardingsphere-agent-core/src/main/java/org/apache/shardingsphere/agent/core/bytebuddy/transformer/ShardingSphereTransformer.java:173]: `IndefinedShape` extends the `MethodDescription`. It's ok for naming as `md` .

File: shardingsphere-agent/shardingsphere-agent-core/src/main/java/org/apache/shardingsphere/agent/core/plugin/PluginApmLoader.java (hunk 8 above)
Comment (dachuan9e) [reply shardingsphere-agent/shardingsphere-agent-core/src/main/java/org/apache/shardingsphere/agent/core/plugin/PluginApmLoader.java:236]: will change

File: shardingsphere-agent/shardingsphere-agent-core/src/main/java/org/apache/shardingsphere/agent/core/bytebuddy/transformer/ShardingSphereTransformer.java (hunk 5 above)
Comment (dachuan9e) [reply shardingsphere-agent/shardingsphere-agent-core/src/main/java/org/apache/shardingsphere/agent/core/bytebuddy/transformer/ShardingSphereTransformer.java]: thanks

**Review by terrymanu — APPROVED** [review]

**Pull request merged**